// @douyinfe/semi-ui/lib/cjs/aiChatInput/extension/inputSlot/component.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("@tiptap/react");
var _react2 = _interopRequireWildcard(require("react"));
var _constants = require("@douyinfe/semi-foundation/lib/cjs/aiChatInput/constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
var _default = props => {
  const isEmpty = props.node.textContent === _constants.strings.ZERO_WIDTH_CHAR;
  const placeholder = props.node.attrs.placeholder || '';
  const placeholderRef = (0, _react2.useRef)(null);
  const [placeholderWidth, setPlaceholderWidth] = (0, _react2.useState)(undefined);
  (0, _react2.useLayoutEffect)(() => {
    if (isEmpty && placeholderRef.current) {
      const timer = setTimeout(() => {
        var _a;
        setPlaceholderWidth((_a = placeholderRef.current) === null || _a === void 0 ? void 0 : _a.offsetWidth);
      });
      return () => {
        clearTimeout(timer);
      };
    }
    return undefined;
  }, [isEmpty, placeholder]);
  return /*#__PURE__*/_react2.default.createElement(_react.NodeViewWrapper, {
    as: "span",
    className: "input-slot",
    style: {
      minWidth: isEmpty && placeholderWidth ? `${placeholderWidth}px` : undefined
    }
  }, isEmpty && (/*#__PURE__*/_react2.default.createElement("span", {
    ref: placeholderRef,
    "data-placeholder": true,
    // contentEditable 必须传入 boolean，strings 导致控制台报错
    // contentEditable' s value must be boolean，strings will cause warning
    contentEditable: false,
    className: "input-slot-placeholder"
  }, placeholder)), /*#__PURE__*/_react2.default.createElement(_react.NodeViewContent, {
    as: "div",
    className: "content"
  }));
};
exports.default = _default;
