// @douyinfe/semi-webpack-plugin/lib/semi-webpack-plugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const utils_1 = require("./utils");
class SemiWebpackPlugin {
    constructor(options) {
        this.options = options;
    }
    apply(compiler) {
        var _a;
        let NormalModule = (_a = this.options.webpackContext) === null || _a === void 0 ? void 0 : _a.NormalModule;
        if (!NormalModule && 'webpack' in compiler)
            NormalModule = compiler.webpack.NormalModule;
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        if (!NormalModule)
            NormalModule = require('webpack/lib/NormalModule');
        compiler.hooks.compilation.tap('SemiPlugin', (compilation) => {
            if (NormalModule.getCompilationHooks) {
                NormalModule.getCompilationHooks(compilation).loader.tap('SemiPlugin', (context, module) => {
                    if (this.options.omitCss) {
                        this.omitCss(module);
                        if (!this.options.webComponentPath) {
                            return;
                        }
                    }
                    this.customTheme(module);
                    if (this.options.prefixCls) {
                        this.customPrefix(module, this.options.prefixCls);
                    }
                    if (this.options.webComponentPath) {
                        this.webComponentAdapter(module);
                    }
                });
            }
            else {
                compilation.hooks.normalModuleLoader.tap('SemiPlugin', (context, module) => {
                    if (this.options.omitCss) {
                        this.omitCss(module);
                        if (!this.options.webComponentPath) {
                            return;
                        }
                    }
                    this.customTheme(module);
                    if (this.options.prefixCls) {
                        this.customPrefix(module, this.options.prefixCls);
                    }
                    if (this.options.webComponentPath) {
                        this.webComponentAdapter(module);
                    }
                });
            }
        });
    }
    webComponentAdapter(module) {
        const compatiblePath = (0, utils_1.transformPath)(module.resource);
        const reg = this.options.webComponentPath instanceof RegExp ? this.options.webComponentPath : /src\/([^/]+\/)*[^/]+\.(ts|tsx|js|jsx)$/;
        if (reg.test(compatiblePath)) {
            module.loaders = module.loaders || [];
            module.loaders.push({
                loader: path_1.default.join(__dirname, 'semi-web-component-loader')
            });
        }
    }
    omitCss(module) {
        const compatiblePath = (0, utils_1.transformPath)(module.resource);
        if (/@douyinfe\/semi-(ui|icons)\/lib\/.+\.js$/.test(compatiblePath)) {
            module.loaders = module.loaders || [];
            module.loaders.push({
                loader: path_1.default.join(__dirname, 'semi-omit-css-loader')
            });
        }
    }
    customTheme(module) {
        var _a, _b, _c;
        const compatiblePath = (0, utils_1.transformPath)(module.resource);
        if (/@douyinfe\/semi-(ui|icons)\/lib\/.+\.js$/.test(compatiblePath)) {
            module.loaders = module.loaders || [];
            module.loaders.push({
                loader: path_1.default.join(__dirname, 'semi-source-suffix-loader')
            });
        }
        if (/@douyinfe\/semi-(ui|icons|foundation)\/lib\/.+\.scss$/.test(compatiblePath)) {
            const scssLoader = require.resolve('sass-loader');
            const cssLoader = require.resolve('css-loader');
            const styleLoader = require.resolve('style-loader');
            const extraCssLoader = path_1.default.join(__dirname, 'semi-extract-css-content-loader');
            const rawLoader = require.resolve('raw-loader');
            const semiSemiLoaderOptions = typeof this.options.theme === 'object' ? Object.assign(Object.assign({}, this.options.theme), { cssLayer: this.options.cssLayer }) : {
                name: this.options.theme,
                cssLayer: this.options.cssLayer
            };
            if (!this.hasSemiThemeLoader(module.loaders)) {
                const lastLoader = this.options.extractCssOptions ? {
                    loader: this.options.extractCssOptions.loader,
                    options: this.options.extractCssOptions.loaderOptions || {}
                } : {
                    loader: styleLoader
                };
                const getRawCssLoaders = [
                    {
                        loader: rawLoader
                    },
                    {
                        loader: extraCssLoader
                    }
                ];
                const commonLoaderList = [
                    {
                        loader: cssLoader,
                        options: {
                            sourceMap: false,
                        }
                    }, {
                        loader: scssLoader,
                        options: {
                            sassOptions: {
                                silenceDeprecations: ['import', 'legacy-js-api', 'global-builtin'],
                            },
                        }
                    },
                    {
                        loader: path_1.default.join(__dirname, 'semi-theme-loader'),
                        options: Object.assign(Object.assign({}, semiSemiLoaderOptions), { prefixCls: this.options.prefixCls, variables: this.convertMapToString(this.options.variables || {}), include: this.options.include })
                    }
                ];
                let loaderList = commonLoaderList;
                if (this.options.webComponentPath) {
                    loaderList = [
                        ...getRawCssLoaders,
                        ...commonLoaderList,
                    ];
                }
                else {
                    loaderList = [
                        lastLoader,
                        ...commonLoaderList,
                    ];
                }
                module.loaders = (_c = (_b = (_a = this.options).overrideStylesheetLoaders) === null || _b === void 0 ? void 0 : _b.call(_a, loaderList)) !== null && _c !== void 0 ? _c : loaderList;
            }
        }
    }
    customPrefix(module, prefix) {
        const compatiblePath = (0, utils_1.transformPath)(module.resource);
        if (/@douyinfe\/semi-[^/]+\/.+env\.js$/.test(compatiblePath)) {
            module.loaders = module.loaders || [];
            module.loaders.push({
                loader: path_1.default.join(__dirname, 'semi-prefix-loader'),
                options: {
                    replacers: {
                        BASE_CLASS_PREFIX: prefix
                    }
                }
            });
        }
    }
    hasSemiThemeLoader(loaders) {
        return (loaders || []).some((loader) => /semi-theme-loader/.test(loader.loader));
    }
    convertMapToString(map) {
        return Object.keys(map).reduce(function (prev, curr) {
            return prev + `${curr}: ${map[curr]};\n`;
        }, '');
    }
}
exports.default = SemiWebpackPlugin;
