// @douyinfe/semi-ui/lib/cjs/aiChatDialogue/Dialogue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var React = _interopRequireWildcard(require("react"));
var _baseComponent = _interopRequireDefault(require("../_base/baseComponent"));
var _classnames = _interopRequireDefault(require("classnames"));
require("@douyinfe/semi-foundation/lib/cjs/aiChatDialogue/aiChatDialogue.css");
var _constants = require("@douyinfe/semi-foundation/lib/cjs/aiChatDialogue/constants");
var _checkbox = require("../checkbox");
var _dialogueTitle = _interopRequireDefault(require("./widgets/dialogueTitle"));
var _dialogueAvatar = _interopRequireDefault(require("./widgets/dialogueAvatar"));
var _dialogueAction = _interopRequireDefault(require("./widgets/dialogueAction"));
var _dialogueContent = _interopRequireDefault(require("./widgets/dialogueContent"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const prefixCls = _constants.cssClasses.PREFIX;
const {
  ROLE,
  DIALOGUE_ALIGN
} = _constants.strings;
class Dialogue extends _baseComponent.default {
  constructor(props) {
    super(props);
    this.getRoleInfo = () => {
      const {
        role,
        message
      } = this.props;
      if (role instanceof Map) {
        return role.get(message === null || message === void 0 ? void 0 : message.name);
      }
      return role;
    };
    this.avatarNode = () => {
      const {
        role,
        dialogueRenderConfig,
        continueSend,
        message
      } = this.props;
      return /*#__PURE__*/React.createElement(_dialogueAvatar.default, {
        role: this.getRoleInfo(),
        message: message,
        customRenderFunc: dialogueRenderConfig === null || dialogueRenderConfig === void 0 ? void 0 : dialogueRenderConfig.renderDialogueAvatar
      });
    };
    this.titleNode = () => {
      const {
        role,
        dialogueRenderConfig
      } = this.props;
      return /*#__PURE__*/React.createElement(_dialogueTitle.default, {
        role: this.getRoleInfo(),
        customRenderFunc: dialogueRenderConfig === null || dialogueRenderConfig === void 0 ? void 0 : dialogueRenderConfig.renderDialogueTitle
      });
    };
    this.contentNode = () => {
      const {
        message,
        mode,
        dialogueRenderConfig,
        markdownRenderProps,
        messageEditRender,
        disabledFileItemClick,
        renderDialogueContentItem,
        onFileClick,
        onImageClick,
        onAnnotationClick,
        onReferenceClick,
        showReference
      } = this.props;
      return /*#__PURE__*/React.createElement(_dialogueContent.default, {
        key: message.editing,
        message: message,
        mode: mode,
        editing: message.editing && message.role === ROLE.USER,
        messageEditRender: messageEditRender,
        onFileClick: onFileClick,
        onImageClick: onImageClick,
        disabledFileItemClick: disabledFileItemClick,
        renderDialogueContentItem: renderDialogueContentItem,
        onAnnotationClick: onAnnotationClick,
        onReferenceClick: onReferenceClick,
        customRenderFunc: dialogueRenderConfig === null || dialogueRenderConfig === void 0 ? void 0 : dialogueRenderConfig.renderDialogueContent,
        markdownRenderProps: markdownRenderProps,
        showReference: showReference
      });
    };
    this.actionNode = () => {
      const {
        role,
        message,
        showReset,
        isLastChat,
        dialogueRenderConfig,
        onMessageReset,
        onMessageGoodFeedback,
        onMessageBadFeedback,
        onMessageCopy,
        onMessageShare,
        messageEditRender,
        onMessageEdit,
        onMessageDelete
      } = this.props;
      return /*#__PURE__*/React.createElement(_dialogueAction.default, {
        role: role,
        message: message,
        onMessageCopy: onMessageCopy,
        onMessageReset: onMessageReset,
        onMessageGoodFeedback: onMessageGoodFeedback,
        onMessageBadFeedback: onMessageBadFeedback,
        showReset: showReset,
        isLastChat: isLastChat,
        onMessageShare: onMessageShare,
        onMessageEdit: onMessageEdit,
        onMessageDelete: onMessageDelete,
        messageEditRender: messageEditRender,
        customRenderFunc: dialogueRenderConfig === null || dialogueRenderConfig === void 0 ? void 0 : dialogueRenderConfig.renderDialogueAction
      });
    };
  }
  render() {
    const {
      message,
      selecting,
      align,
      isSelected,
      onSelectChange,
      continueSend
    } = this.props;
    const id = message.id;
    const isRightAlign = message.role === ROLE.USER && align === DIALOGUE_ALIGN.LEFT_RIGHT;
    const containerCls = (0, _classnames.default)({
      [`${prefixCls}-container`]: true,
      [`${prefixCls}-container-right`]: isRightAlign
    });
    return /*#__PURE__*/React.createElement("div", {
      className: (0, _classnames.default)(`${prefixCls}-wrapper`, {
        [`${prefixCls}-wrapper-selected`]: selecting && isSelected,
        [`${prefixCls}-wrapper-continue-send`]: continueSend
      })
    }, selecting && (/*#__PURE__*/React.createElement("div", {
      className: `${prefixCls}-checkbox`
    }, /*#__PURE__*/React.createElement(_checkbox.Checkbox, {
      checked: isSelected,
      onChange: e => onSelectChange(e.target.checked, id)
    }))), /*#__PURE__*/React.createElement("div", {
      className: containerCls
    }, this.avatarNode(), /*#__PURE__*/React.createElement("div", {
      className: `${prefixCls}-inner`
    }, !continueSend && this.titleNode(), this.contentNode(), this.actionNode())));
  }
}
var _default = exports.default = Dialogue;
