// @douyinfe/semi-ui/lib/es/aiChatInput/richTextInput.js
import { EditorContent, useEditor } from '@tiptap/react';
import React, { useCallback, useEffect, useMemo, useRef } from 'react';
import Document from '@tiptap/extension-document';
import Text from '@tiptap/extension-text';
import { UndoRedo } from '@tiptap/extensions';
import Paragraph from '@tiptap/extension-paragraph';
import HardBreak from '@tiptap/extension-hard-break';
import { Placeholder } from '@tiptap/extensions';
import InputSlot from './extension/inputSlot';
import SelectSlot from './extension/selectSlot';
import SkillSlot from './extension/skillSlot';
import { cssClasses } from '@douyinfe/semi-foundation/lib/es/aiChatInput/constants';
import { handleCompositionEndLogic, handlePasteLogic, handleTextInputLogic, handleZeroWidthCharLogic } from './extension/plugins';
import SemiStatusExtension from './extension/statusExtension';
const PREFIX = cssClasses.PREFIX;
export default props => {
  const {
    setEditor,
    onKeyDown,
    onChange,
    placeholder,
    extensions = [],
    defaultContent,
    onPaste,
    innerRef,
    handleKeyDown,
    onFocus,
    onBlur,
    handleCreate
  } = props;
  const isComposing = useRef(false);
  const handleCompositionStart = useCallback(view => {
    isComposing.current = true;
  }, []);
  const handleCompositionEnd = useCallback(view => {
    isComposing.current = false;
    handleCompositionEndLogic(view);
  }, []);
  const handleTextInput = useCallback((view, from, to, text) => {
    if (isComposing.current) {
      return false;
    }
    return handleTextInputLogic(view, from, to, text);
  }, []);
  const allExtensions = useMemo(() => {
    return [Document, Paragraph, Text, UndoRedo, HardBreak, InputSlot, SelectSlot, SkillSlot, Placeholder.configure({
      placeholder: placeholder
    }), SemiStatusExtension, ...extensions];
  }, [extensions, placeholder]);
  const editorProps = useMemo(() => {
    return {
      handleKeyDown: handleKeyDown,
      handlePaste: handlePasteLogic,
      handleTextInput,
      handleDOMEvents: {
        compositionstart: handleCompositionStart,
        compositionend: handleCompositionEnd
      }
    };
  }, [handleKeyDown, handleTextInput, handleCompositionStart, handleCompositionEnd]);
  // const onSelectionUpdate = useCallback(({ editor }) => {
  //     // For debug
  //     const fromPos = editor.state.selection.from;
  //     const { $from } = editor.state.selection;
  //     console.log('光标/选区位置', fromPos, editor.state.selection, editor.state.doc);
  //     // console.log('before', $from.nodeBefore, $from.nodeAfter);
  // }, []);
  const onCreate = useCallback(_ref => {
    let {
      editor
    } = _ref;
    const {
      state,
      view
    } = editor;
    const tr = handleZeroWidthCharLogic(state);
    if (tr) {
      // 一次性触发，避免多次触发导致 appendTransaction 被多次调用
      view.dispatch(tr);
    }
    handleCreate();
  }, [handleCreate]);
  const onUpdate = useCallback(_ref2 => {
    let {
      editor
    } = _ref2;
    // The content has changed.
    const content = editor.getText();
    onChange(content);
  }, [onChange]);
  const handlePaste = useCallback(e => {
    var _a;
    // To support file paste
    const items = (_a = e.clipboardData) === null || _a === void 0 ? void 0 : _a.items;
    let files = [];
    if (items) {
      for (const it of items) {
        const file = it.getAsFile();
        file && files.push(it.getAsFile());
      }
    }
    if (files.length) {
      onPaste === null || onPaste === void 0 ? void 0 : onPaste(files);
    }
  }, [onPaste]);
  const editor = useEditor({
    extensions: allExtensions,
    content: defaultContent !== null && defaultContent !== void 0 ? defaultContent : ``,
    editorProps: editorProps,
    // onSelectionUpdate,
    onCreate,
    onUpdate,
    onPaste: handlePaste
  });
  useEffect(() => {
    setEditor(editor);
  }, [editor, setEditor]);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(EditorContent, {
    editor: editor,
    onKeyDown: onKeyDown,
    onFocus: onFocus,
    onBlur: onBlur,
    ref: innerRef,
    className: `${PREFIX}-editor-content`
  }));
};
