// @douyinfe/semi-foundation/lib/cjs/aiChatDialogue/dataAdapter/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "chatCompletionToMessage", {
  enumerable: true,
  get: function () {
    return _chatCompletionToMessage.default;
  }
});
Object.defineProperty(exports, "chatInputToChatCompletion", {
  enumerable: true,
  get: function () {
    return _chatInputToChatCompletion.default;
  }
});
Object.defineProperty(exports, "chatInputToMessage", {
  enumerable: true,
  get: function () {
    return _chatInputToMessage.default;
  }
});
Object.defineProperty(exports, "messageToChatInput", {
  enumerable: true,
  get: function () {
    return _messageToChatInput.default;
  }
});
Object.defineProperty(exports, "responseToMessage", {
  enumerable: true,
  get: function () {
    return _responseToMessage.default;
  }
});
Object.defineProperty(exports, "streamingChatCompletionToMessage", {
  enumerable: true,
  get: function () {
    return _streamingChatCompletionToMessage.default;
  }
});
Object.defineProperty(exports, "streamingResponseToMessage", {
  enumerable: true,
  get: function () {
    return _streamingResponseToMessage.default;
  }
});
var _responseToMessage = _interopRequireDefault(require("./responseToMessage"));
var _chatCompletionToMessage = _interopRequireDefault(require("./chatCompletionToMessage"));
var _streamingChatCompletionToMessage = _interopRequireDefault(require("./streamingChatCompletionToMessage"));
var _streamingResponseToMessage = _interopRequireDefault(require("./streamingResponseToMessage"));
var _chatInputToMessage = _interopRequireDefault(require("./chatInputToMessage"));
var _chatInputToChatCompletion = _interopRequireDefault(require("./chatInputToChatCompletion"));
var _messageToChatInput = _interopRequireDefault(require("./messageToChatInput"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
